From cee72e388eac2e86a694cece68bf44aedc955665 Mon Sep 17 00:00:00 2001
From: "GONG, Ruiqi" <gongruiqi1@huawei.com>
Date: Tue, 31 May 2022 05:33:18 +0800
Subject: [PATCH 2186/2592] [Huawei] selinux: genfscon: fix a potential
 out-of-bound write in sel_write_ns

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186870

----------------------------------------

It was found in sel_write_ns() that an out-of-bound write could happen
caused by a deliberately constructed input from user space that makes
strlen(page) == 0. For example:

  echo -ne "\0..." > /sys/fs/selinux/ns
  printf "\0..." > /sys/fs/selinux/ns

This patch refactors the way how sel_write_ns() deals with input by
making use of sscanf() and imitating how other selinuxfs interfaces do.

Fixes: 5f4300d116cf ("[Huawei] selinux: genfscon ns: Enable processes to switch namespace")
Signed-off-by: GONG, Ruiqi <gongruiqi1@huawei.com>
---
 security/selinux/selinuxfs.c | 32 +++++++++++++++++---------------
 1 file changed, 17 insertions(+), 15 deletions(-)

diff --git a/security/selinux/selinuxfs.c b/security/selinux/selinuxfs.c
index f24aa04ccd20..2ba2da4461e7 100644
--- a/security/selinux/selinuxfs.c
+++ b/security/selinux/selinuxfs.c
@@ -337,40 +337,41 @@ static const struct file_operations sel_disable_ops = {
 static ssize_t sel_write_ns(struct file *file, const char __user *buf,
 			    size_t count, loff_t *ppos)
 {
-	char *page, *ptr;
+	char *page = NULL, *name = NULL, *ptr;
 	char c;
 	ssize_t length;
 	struct selinux_ns *new_ns;
 
 	if (count >= PAGE_SIZE)
 		return -ENOMEM;
+	/* Not support ns nesting or switching */
+	if (current_selinux_ns->parent != NULL)
+		return -EPERM;
 
 	page = memdup_user_nul(buf, count);
 	if (IS_ERR(page))
 		return PTR_ERR(page);
-	/* strip any trailing newline */
-	if (page[strlen(page) - 1] == '\n')
-		page[strlen(page) - 1] = 0;
-	/* cannot be the same name as init_ns */
-	if (!strcmp(page, SELINUX_NS_INIT_NAME)) {
-		length = -EINVAL;
+
+	length = -ENOMEM;
+	name = kzalloc(count + 1, GFP_KERNEL);
+	if (!name)
 		goto err_out;
-	}
-	/* Not support ns nesting or switching */
-	if (current_selinux_ns->parent != NULL) {
-		length = -EPERM;
+
+	length = -EINVAL;
+	if (sscanf(page, "%s", name) != 1)
+		goto err_out;
+	/* cannot be the same name as init_ns */
+	if (!strcmp(name, SELINUX_NS_INIT_NAME))
 		goto err_out;
-	}
 	/* only allow [0-9A-Za-z_] in ns's name */
-	for (ptr = page; *ptr != '\0'; ++ptr) {
+	for (ptr = name; *ptr != '\0'; ++ptr) {
 		c = *ptr;
 		if (isalnum(c) || (c == '_'))
 			continue;
-		length = -EINVAL;
 		goto err_out;
 	}
 
-	new_ns = selinux_ns_create(page);
+	new_ns = selinux_ns_create(name);
 	if (IS_ERR(new_ns)) {
 		length = PTR_ERR(new_ns);
 		goto err_out;
@@ -386,6 +387,7 @@ static ssize_t sel_write_ns(struct file *file, const char __user *buf,
 	length = count;
 err_out:
 	kfree(page);
+	kfree(name);
 	return length;
 }
 
-- 
2.17.1

